File: web/decisions.js
"use strict";
Object.assign(titles, {decisions:["Decision inbox", "Choose a direction here. The brain continues within the recorded scope."]});
const decisionDrafts = new Map();
const decisionDetailsOpen = new Set();
const decisionLabels = {open:"Needs your decision", answered:"Answer recorded", received:"Received by brain", applied:"Applied to design", blocked:"Needs follow-up", superseded:"Superseded"};

function workflowSummary(root, controls=false) {
  const w=state.workflow;
  if(!w) {root.append(callout("Workflow upgrade needs a server restart", "No listener or decision state is available from this running server.")); return;}
  const panel=el("section",null,"workflow-summary");
  const label={off:"Idle listening is off",needs_activation:"Native listener needs activation",unconfirmed:"Listener schedule recorded · check-in overdue or missing",listening:"Listener checked in recently"}[w.status];
  panel.append(el("p","DECISIONS & CONTINUATION","eyebrow"),el("h2",label));
  panel.append(el("p",`${w.openDecisions} awaiting your decision · ${w.pendingRequests} requests awaiting completion · Worker dispatch ${w.dispatchPaused?'paused':'enabled'}`));
  panel.append(el("p",`Every 15 minutes while enabled · Inbox checked ${when(w.lastCheckedAt)}`,"muted"));
  const info=el("details",null,"coverage-details"),infoKey='listener:'+controls;
  info.open=decisionDetailsOpen.has(infoKey);
  info.addEventListener('toggle',()=>{if(info.open)decisionDetailsOpen.add(infoKey);else decisionDetailsOpen.delete(infoKey);});
  info.append(el("summary","Schedule, usage & availability"));
  info.append(el("p",`Native heartbeat: ${w.nativeStatus} (observed ${when(w.nativeObservedAt)}).`));
  info.append(el("p","Keep this computer and Codex running. Scheduled checks consume model usage, including idle checks; this is not an instant browser-to-Codex connection."));
  const actions=el("div",null,"inline-actions");
  if(!controls) actions.append(button(w.openDecisions?`Review ${w.openDecisions} decision${w.openDecisions===1?'':'s'}`:"Open decision inbox",()=>navigateView("decisions"),w.openDecisions?"primary":""));
  else {
    actions.append(button(w.listenerEnabled?"Turn off idle listening":"Keep listening between jobs",()=>command("listening",{enabled:!w.listenerEnabled})));
    info.append(el("p","Listening and worker dispatch are separate. Turning off idle listening takes effect on the next brain cycle; active workers still need supervision. Turning it back on cannot wake an already-paused native schedule."));
  }
  if(w.status==='needs_activation'||w.status==='unconfirmed') {
    panel.append(el("p",w.status==='needs_activation'?"Open the brain once to activate its native decision listener. A saved request cannot wake a paused schedule.":"No fresh inbox check or native status confirmation. Check the brain and app availability; the saved schedule alone does not prove it is running.","muted"));
    if(/^[a-zA-Z0-9_-]{1,100}$/.test(state.meta.brainId||"")) {
      const link=el("a","Open brain in Codex","button");link.href="codex://threads/"+encodeURIComponent(state.meta.brainId);actions.append(link);
    }
  }
  panel.append(actions,info);root.append(panel);
}

function decisionArtifacts(root, ids) {
  const links=el("div",null,"inline-actions");
  for(const id of ids) {
    const artifact=state.observations.artifacts.find(a=>a.id===id);
    if(artifact) links.append(button(`Read ${artifact.name} · v${artifact.version}`,()=>{
      observationRepo="all";artifactQuery="";navigateView("artifacts",id);
      document.querySelector('.artifact-reader')?.scrollIntoView();
    }));
  }
  root.append(links);
}

async function submitDecision(d, draft, submit) {
  if(busy||!connected) {showNotice("Wait for the current request or refresh the connection.",true);return;}
  const payload={decisionId:d.id,decisionHash:d.decisionHash,optionId:draft.optionId,note:draft.note,confirmed:draft.confirmed};
  // Retain the exact envelope after an uncertain network response; retry is not a second answer.
  if(!draft.request||JSON.stringify(draft.request.payload)!==JSON.stringify(payload))
    draft.request={id:crypto.randomUUID(),kind:"decision_response",expectedRevision:state.meta.revision,payload};
  busy=true;submit.disabled=true;
  try {
    await api('/api/commands',{method:'POST',headers:{'Content-Type':'application/json','X-CSRF-Token':csrf},body:JSON.stringify(draft.request)});
    decisionDrafts.delete(d.id);
    showNotice("Answer recorded for this exact version. The brain will receive it on its next active cycle; no implementation was authorized.");
    await refresh();
  } catch(error) {
    if(error.message.includes("State changed")) draft.request=null;
    showNotice(error.message+" Your draft is retained. Refresh and review before retrying.",true);
  } finally {busy=false;submit.disabled=false;}
}

function decisionCard(d, root) {
  const card=el("article",null,"decision-card"),s=d.spec;
  card.append(el("p",`${s.repository} / ${s.key} · version ${d.version}`,"eyebrow"),el("h2",s.title),badge(decisionLabels[d.status]||d.status));
  card.append(el("h3",s.question),el("p",s.context));
  card.append(el("p","Scope: "+s.scope,"decision-scope"),el("p","Next step: "+s.nextStep),el("p",d.boundary,"muted"));
  decisionArtifacts(card,s.artifactIds);
  const detail=el("details",null,"coverage-details");
  detail.open=decisionDetailsOpen.has('binding:'+d.id);
  detail.addEventListener('toggle',()=>{if(detail.open)decisionDetailsOpen.add('binding:'+d.id);else decisionDetailsOpen.delete('binding:'+d.id);});
  detail.append(el("summary","Version binding & creation time"),el("p",d.decisionHash,"mono"),el("p",when(d.createdAt)));card.append(detail);
  if(d.status==='open') {
    let draft=decisionDrafts.get(d.id);
    if(!draft){draft={optionId:"",note:"",confirmed:false,request:null};decisionDrafts.set(d.id,draft);}
    const form=el("form",null,"decision-form"),choices=el("fieldset");choices.append(el("legend","Choose a direction"));
    s.options.forEach(option=>{
      const label=el("label",null,"decision-option"),radio=el("input"),copy=el("span");
      radio.type="radio";radio.name="decision-"+d.id;radio.value=option.id;radio.checked=draft.optionId===option.id;radio.required=true;
      copy.append(el("strong",option.label+(option.id===s.recommendedOptionId?" · Recommended":"")),el("span",option.implications,"subline"));
      label.append(radio,copy);choices.append(label);
      radio.onchange=()=>{draft.optionId=option.id;draft.confirmed=false;check.checked=false;note.required=option.requiresNote;};
    });
    const noteLabel=el("label", "Your note or requested information"),note=el("textarea");note.rows=4;note.maxLength=4000;note.value=draft.note;
    note.placeholder="No credentials or secrets. This note is input for review, not permission to execute commands.";
    note.required=s.options.find(o=>o.id===draft.optionId)?.requiresNote||false;
    note.oninput=()=>{draft.note=note.value;draft.confirmed=false;check.checked=false;};noteLabel.append(note);
    const confirmation=el("label",null,"decision-confirmation"),check=el("input");check.type="checkbox";check.required=true;check.checked=draft.confirmed;
    check.onchange=()=>{draft.confirmed=check.checked;};confirmation.append(check,el("span","I confirm this direction and note for this version. This is not approval to implement, access targets or merge."));
    const submit=button("Record decision",()=>{},"primary");submit.type="submit";
    form.onsubmit=e=>{e.preventDefault();if(form.reportValidity())submitDecision(d,draft,submit);};
    form.append(choices,noteLabel,confirmation,submit);card.append(form);
  } else if(d.response) {
    const option=s.options.find(o=>o.id===d.response.optionId);
    card.append(el("p","Your answer: "+option.label),el("p",d.response.note||"No additional note.","decision-note"),el("p","Recorded "+when(d.response.at),"muted"));
    if(d.receivedAt)card.append(el("p","Received by brain "+when(d.receivedAt),"muted"));
  }
  if(d.resolution){card.append(el("h3","Brain outcome"),el("p",d.resolution.summary),el("p",when(d.resolution.at),"muted"));decisionArtifacts(card,d.resolution.artifactIds);}
  root.append(card);
}

function decisions(root) {
  workflowSummary(root,true);
  const rows=state.decisions||[],active=rows.filter(d=>['open','answered','received'].includes(d.status));
  if(!active.length)root.append(empty("No decisions waiting", "The brain publishes specific questions here with their scope and supporting artifacts. Already answered decisions stay in history. Packet execution still needs a separate approval in Approved queue."));
  active.sort((a,b)=>a.createdAt-b.createdAt).forEach(d=>decisionCard(d,root));
  const history=rows.filter(d=>!['open','answered','received'].includes(d.status));
  if(history.length){root.append(section("Decision history","Earlier versions and outcomes remain preserved."));history.forEach(d=>{const details=el("details",null,"decision-history");details.open=decisionDetailsOpen.has(d.id);details.addEventListener('toggle',()=>{if(details.open)decisionDetailsOpen.add(d.id);else decisionDetailsOpen.delete(d.id);});details.append(el("summary",`${d.spec.title} · v${d.version} · ${decisionLabels[d.status]}`));decisionCard(d,details);root.append(details);});}
}

File: orchestrator/decisions.py
"""Version-bound owner decisions. Data and receipts, never execution authority."""
import re
import time

from .core import ACTIVE, SHA, canonical, digest, require

FIELDS = {"key", "repository", "title", "question", "context", "scope", "nextStep",
          "options", "recommendedOptionId", "artifactIds"}
BOUNDARY = "Design/input only. No packet approval, dispatch, target access, privilege, merge or acceptance authority."


def text(value, name, limit=4000, empty=False):
    require(isinstance(value, str) and len(value) <= limit and (empty or value.strip()), "Invalid " + name)


def artifacts(ledger, db, ids, repository):
    require(isinstance(ids, list) and 1 <= len(ids) <= 8 and all(isinstance(i, str) and SHA.fullmatch(i) for i in ids), "Use 1–8 retained artifact IDs")
    require(len(set(ids)) == len(ids), "Duplicate artifact reference")
    for identity in ids:
        item = ledger.get(db, "artifact_versions", identity)
        require(item["repository"] == repository, "Artifact belongs to another repository")


def authorize_brain(ledger, db, token):
    meta = ledger.authorize(db, token)
    owner = meta["controller"]["owner"]
    require(meta["brainId"] and (owner == meta["brainId"] or owner.startswith(meta["brainId"] + ":")), "Only the designated brain can publish or resolve decisions")
    return meta


def publish(ledger, token, spec):
    require(isinstance(spec, dict) and set(spec) == FIELDS, "Decision fields must match the decision contract")
    for name in ("key", "repository", "title", "question", "context", "scope", "nextStep"):
        text(spec[name], name, 200 if name in ("key", "repository", "title") else 4000)
    require(re.fullmatch(r"[A-Za-z0-9][A-Za-z0-9_.-]{0,99}", spec["key"]), "Invalid decision key")
    options = spec["options"]
    require(isinstance(options, list) and 2 <= len(options) <= 5, "Provide 2–5 bounded options")
    for option in options:
        require(isinstance(option, dict) and set(option) == {"id", "label", "implications", "requiresNote"}, "Invalid option fields")
        text(option["id"], "option ID", 64)
        require(re.fullmatch(r"[a-z0-9][a-z0-9_-]*", option["id"]), "Invalid option ID")
        text(option["label"], "option label", 200)
        text(option["implications"], "option implications", 2000)
        require(type(option["requiresNote"]) is bool, "requiresNote must be boolean")
    ids = [o["id"] for o in options]
    require(len(ids) == len(set(ids)) and spec["recommendedOptionId"] in [None, *ids], "Invalid recommendation or duplicate option")
    with ledger.tx() as db:
        meta = authorize_brain(ledger, db, token)
        ledger.get(db, "repos", spec["repository"])
        artifacts(ledger, db, spec["artifactIds"], spec["repository"])
        previous = sorted((d for d in ledger.all(db, "decisions") if d["spec"]["repository"] == spec["repository"] and d["spec"]["key"] == spec["key"]), key=lambda d: d["version"])
        if previous and previous[-1]["spec"] == spec:
            return previous[-1]  # Publishing twice never reopens an answered decision.
        if previous:
            old = previous[-1]
            require(old["status"] != "received", "Decision is in flight; reconcile and resolve before revising")
            if old["status"] in ("open", "answered"):
                old["status"] = "superseded"
                ledger.put(db, "decisions", old["id"], old)
                if old["response"]:
                    cmd = ledger.get(db, "commands", old["response"]["commandId"])
                    cmd.update(status="rejected", result="Decision revised; review the new version before answering.")
                    ledger.put(db, "commands", cmd["id"], cmd)
        version = len(previous) + 1
        document = {"schemaVersion": 1, "version": version, "brainId": meta["brainId"], "spec": spec, "boundary": BOUNDARY}
        identity = digest(document)
        record = {**document, "id": identity, "decisionHash": identity, "status": "open", "createdAt": time.time(), "response": None, "resolution": None}
        db.execute("INSERT INTO snapshots VALUES (?,?,?)", (identity, "decision", canonical(document)))
        ledger.put(db, "decisions", identity, record)
        ledger.event(db, "decision_published", {"id": identity, "key": spec["key"], "version": version})
        return record


def answer(ledger, db, command):
    p = command["payload"]
    require(set(p) == {"decisionId", "decisionHash", "optionId", "note", "confirmed"}, "Invalid decision response fields")
    for name in ("decisionId", "decisionHash"):
        text(p[name], name, 128)
    if p["optionId"] is not None:
        text(p["optionId"], "option ID", 64)
    text(p["note"], "decision note", 4000, empty=True)
    require(p["confirmed"] is True, "Explicit decision confirmation required")
    d = ledger.get(db, "decisions", p["decisionId"])
    require(d["status"] == "open" and d["decisionHash"] == p["decisionHash"], "Decision changed or already answered; review its current version")
    if p["optionId"] is None:
        require(bool(p["note"].strip()), "Write your answer or choose a suggested option")
    else:
        option = next((o for o in d["spec"]["options"] if o["id"] == p["optionId"]), None)
        require(option is not None, "Unknown decision option")
        require(not option["requiresNote"] or bool(p["note"].strip()), "This option needs the requested information in your note")
    d.update(status="answered", response={"commandId": command["id"], "answerKind": "free_text" if p["optionId"] is None else "option",
                                         "optionId": p["optionId"], "note": p["note"], "at": time.time()})
    ledger.put(db, "decisions", d["id"], d)


def receive(ledger, db, cmd):
    d = ledger.get(db, "decisions", cmd["payload"]["decisionId"])
    require(d["status"] == "answered" and d["response"]["commandId"] == cmd["id"], "Decision response is no longer current")
    d.update(status="received", receivedAt=time.time())
    ledger.put(db, "decisions", d["id"], d)
    cmd.update(status="processing", result="Received by the brain; continuation and evidence still pending.")
    return {"kind": "decision_response", "commandId": cmd["id"], "decision": d, "boundary": BOUNDARY}


def resolve(ledger, token, identity, result):
    require(isinstance(result, dict) and set(result) == {"commandId", "outcome", "summary", "artifactIds"}, "Invalid decision resolution")
    require(result["outcome"] in ("applied", "blocked"), "Outcome must be applied or blocked")
    text(result["summary"], "resolution summary")
    text(result["commandId"], "response command ID", 128)
    with ledger.tx() as db:
        authorize_brain(ledger, db, token)
        d = ledger.get(db, "decisions", identity)
        if d["resolution"]:
            require({k: d["resolution"][k] for k in result} == result, "Resolution already recorded differently")
            return d
        require(d["status"] == "received" and d["response"]["commandId"] == result["commandId"], "Exact received response required")
        artifacts(ledger, db, result["artifactIds"], d["spec"]["repository"])
        cmd = ledger.get(db, "commands", result["commandId"])
        require(cmd["status"] == "processing", "Response is not in flight")
        d.update(status=result["outcome"], resolution={**result, "at": time.time()})
        ledger.put(db, "decisions", identity, d)
        cmd.update(status="completed", result=result["summary"])
        ledger.put(db, "commands", cmd["id"], cmd)
        ledger.event(db, "decision_resolved", {"id": identity, "outcome": result["outcome"], "artifactIds": result["artifactIds"]})
        return d


def workflow(state):
    """Requested policy and observed native state are deliberately separate."""
    m, now = state["meta"], state["serverTime"]
    enabled = m.get("decisionListener", {}).get("enabled", False)
    heartbeat = m["heartbeat"]
    checked = m.get("inboxCheckedAt")
    observed = heartbeat.get("observedAt")
    recent = (checked is not None and 0 <= now - checked <= 35 * 60
              and observed is not None and 0 <= now - observed <= 35 * 60)
    pending = [c for c in state["commands"] if c["status"] in ("queued", "processing") or c.get("needsBrainReceipt")]
    active = any(w["status"] in ACTIVE for w in state["workers"])
    approved = not m["paused"] and not state.get("admission", {}).get("dispatchBlocked") and any(q["status"] == "approved" and not q["held"] for q in state["queue"])
    continuations = state.get("continuations", [])
    unplanned = sum(c["status"] in ("needs_proposal", "needs_revision") for c in continuations)
    reasons = []
    for needed, reason in ((active, "active_workers"), (bool(m.get("runner")), "runner_owned"),
                           (approved, "approved_dispatch"), (bool(pending), "pending_receipts"),
                           (bool(unplanned), "follow_up_planning")):
        if needed:
            reasons.append(reason)
    keep = enabled or bool(reasons)
    status = ("needs_activation" if heartbeat["status"] != "ACTIVE" else "listening" if recent else "unconfirmed") if keep else (
        "idle_pause_pending" if heartbeat["status"] == "ACTIVE" else "event_waiting")
    brain = m.get("brainControl", {})
    parked = brain.get("phase") == "parked"
    if parked:
        status = "brain_stopped"
    return {"listenerEnabled": enabled, "status": status, "lastCheckedAt": checked,
            "nativeStatus": heartbeat["status"], "nativeObservedAt": heartbeat.get("observedAt"),
            "intervalMinutes": 15, "pendingRequests": len(pending),
            "openDecisions": sum(d["status"] == "open" for d in state["decisions"]),
            "followUpsNeedingProposal": unplanned, "followUps": len(continuations),
            "schedulingMode": "periodic_idle" if enabled else "event_driven",
            "supervisionReasons": reasons, "shouldKeepHeartbeat": not parked and keep,
            "dispatchPaused": m["paused"], "boundary": BOUNDARY}


def inbox(state):
    """Compact native-cycle input: no token logs, artifact bodies or event history."""
    return {"meta":state["meta"], "workflow":state["workflow"], "continuations": state.get("continuations", []),
            "admission": state.get("admission"),
            "missionConfiguration": {k: state.get("mission", {}).get(k) for k in ("version", "effectiveStatus", "documentHash", "activation", "executionAuthority")},
            "decisions":[d for d in state["decisions"] if d["status"] in ("open", "answered", "received")],
            "commands":[c for c in state["commands"] if c["status"] in ("queued", "processing") or c.get("needsBrainReceipt")],
            "queue":[q for q in state["queue"] if q["status"] == "approved"],
            "workers":[w for w in state["workers"] if w["status"] in ACTIVE]}
